File: bindings/dart/lib/src/vector_layer.dart
// lib/src/vector_layer.dart
//
// VectorLayer: approximate-nearest-neighbour index on top of WaveDB.
//
// Mirrors graph_layer.dart's wrapper pattern: hold the opaque ffi pointer,
// call lookupFunction via WaveDBNative, encode strings/bytes, check rc, map
// errors, lifecycle via dispose() + NativeFinalizer.
//
// Config is split into a Format tier (immutable after construct) and a
// Runtime tier (freely mutable via reconfigure). The wrapper exposes these
// as `VectorFormat` and `VectorRuntime` classes; `_fillFormat`/`_fillRuntime`
// copy them into the C structs.
//
// Only the sync API is wired here (insert/search/delete/train/rebuild/count/
// reconfigure). The async variants take a caller-owned `promise_t*` whose
// resolve/reject callbacks would need a NativeCallable.listener trampoline
// (the graph_layer.dart pattern). Declared in wavedb_bindings.dart so the
// symbols resolve, but VectorLayer does not call them — deferred for a
// follow-up. The sync API is the priority.
import 'dart:ffi';
import 'dart:typed_data';
import 'package:ffi/ffi.dart';

import 'native/types.dart';
import 'native/wavedb_bindings.dart';
import 'exceptions.dart';
import 'subtree.dart';
import 'database.dart';

/// C `vl_index_type_t` enum values.
class IndexType {
  const IndexType._();
  static const int flat = 0; // exact brute-force; baseline + IVF cold-start
  static const int ivf = 1; // inverted file
  static const int slsh = 2; // SK-LSH bidirectional scan
}

/// C `vl_distance_t` enum values.
class Distance {
  const Distance._();
  static const int l2 = 0;
  static const int cosine = 1; // default for embedding workloads
  static const int dot = 2;
}

/// Immutable format tier — set at create time, cannot be changed after.
///
/// To change any field here, drop the layer (and its subtree / separate db)
/// and recreate. `vector_layer_reconfigure` accepts a [VectorRuntime] only;
/// the C signature makes format mutation structurally impossible.
class VectorFormat {
  final int indexType;
  final int dim;
  final String delimiter;
  final int distance;
  final int ivfNClusters;
  final int slshLshTables;
  final int slshHashBits;
  final double slshBucketWidth;

  const VectorFormat({
    required this.indexType,
    required this.dim,
    this.delimiter = '/',
    this.distance = Distance.cosine,
    this.ivfNClusters = 0,
    this.slshLshTables = 0,
    this.slshHashBits = 0,
    this.slshBucketWidth = 0.0,
  });
}

/// Runtime tier — freely mutable via [VectorLayer.reconfigure].
class VectorRuntime {
  final int topK;
  final int syncOnly;
  final int ivfNprobe;
  final int ivfFlatUntil;
  final int slshScanRadius;

  const VectorRuntime({
    this.topK = 10,
    this.syncOnly = 1,
    this.ivfNprobe = 8,
    this.ivfFlatUntil = 1000,
    this.slshScanRadius = 200,
  });
}

/// One hit from [VectorLayer.searchSync].
///
/// `id` is the raw string passed to [VectorLayer.insertSync]. `metadata` is
/// the raw bytes passed to `insertSync` (empty `Uint8List` if none).
/// `distance` is the metric distance (lower = closer for L2; for cosine/dot
/// the C layer reports the post-transform value — see
/// src/Layers/vector/vector_distance.c).
class VectorResult {
  final String id;
  final double distance;
  final Uint8List metadata;

  const VectorResult(this.id, this.distance, this.metadata);

  String get metadataAsString =>
      String.fromCharCodes(metadata);

  @override
  String toString() =>
      'VectorResult(id=$id, distance=${distance.toStringAsFixed(4)}, '
      'metadata=${metadata.length} bytes)';
}

/// Exception thrown by vector layer operations.
class VectorLayerException implements Exception {
  final String message;
  const VectorLayerException(this.message);

  @override
  String toString() => 'VectorLayerException: $message';
}

/// ANN index layered on a WaveDB database (shared) or a dedicated db.
///
/// Two construction modes:
///
/// - [VectorLayer.openSeparate] — dedicated database at `dbLocation`. Clean
///   LRU/WAL tuning, no key-space sharing. Mirrors the Hippo document_db
///   split.
///
/// - [VectorLayer.open] — shares `db`'s underlying database. If `subtree`
///   is supplied, all keys land under that subtree prefix; otherwise keys
///   land directly in `db` under `{indexName}/...`.
class VectorLayer implements Finalizable {
  Pointer<vector_layer_t>? _vl;
  bool _closed = false;
  // Keep a reference to the parent db/subtree so they aren't GC'd while
  // the layer is alive. The C vector_layer_t does NOT hold a refcount on
  // the db/subtree, so the Dart side must. Accessed via [dbRef]/[subtreeRef]
  // below to keep the analyzer from flagging these as unused.
  final WaveDB? _db;
  final Subtree? _subtree;

  /// Reference to the parent database (for shared-db mode); null when opened
  /// via [openSeparate]. Held to prevent the parent [WaveDB] from being
  /// garbage-collected while this layer is alive.
  WaveDB? get dbRef => _db;

  /// Reference to the parent subtree, if any. Held to prevent GC of the
  /// [Subtree] while this layer is alive.
  Subtree? get subtreeRef => _subtree;

  static final NativeFinalizer _finalizer = NativeFinalizer(
    WaveDBNative.vectorLayerDestroyNative.cast(),
  );

  VectorLayer._(this._vl, this._db, this._subtree) {
    _finalizer.attach(this, _vl!.cast(), detach: this);
  }

  // ── construction ──

  /// Open a vector layer on a DEDICATED database at `dbLocation`.
  factory VectorLayer.openSeparate(
    String dbLocation,
    String indexName,
    VectorFormat fmt,
    VectorRuntime rt,
  ) {
    if (indexName.isEmpty) {
      throw VectorLayerException('indexName must be non-empty');
    }
    final cfg = _newConfig(fmt, rt);
    try {
      final ptr = WaveDBNative.vectorLayerOpenSeparate(
          dbLocation, indexName, cfg);
      return VectorLayer._(ptr, null, null);
    } finally {
      calloc.free(cfg);
    }
  }

  /// Open a vector layer sharing `db`'s underlying database. If `subtree`
  /// is supplied, all keys land under that subtree prefix.
  factory VectorLayer.open(
    String indexName,
    WaveDB db,
    VectorFormat fmt,
    VectorRuntime rt, {
    Subtree? subtree,
  }) {
    if (indexName.isEmpty) {
      throw VectorLayerException('indexName must be non-empty');
    }
    final cfg = _newConfig(fmt, rt);
    try {
      final ptr = WaveDBNative.vectorLayerCreate(
        indexName,
        db.databasePtr,
        subtree?.nativePtr ?? nullptr,
        cfg,
      );
      return VectorLayer._(ptr, db, subtree);
    } finally {
      calloc.free(cfg);
    }
  }

  // ── sync API ──

  /// Insert `vec` under `id`, with optional `metadata` bytes. Returns the C
  /// rc (0 on success, <0 on failure).
  int insertSync(String id, List<double> vec, {List<int>? metadata}) {
    _checkOpen();
    final meta = metadata ?? const <int>[];
    final rc = WaveDBNative.vectorLayerInsertSync(_vl!, id, vec, meta);
    if (rc < 0) {
      throw WaveDBException.ioError(
          'vector_layer_insert_sync', 'Insert failed (rc=$rc)');
    }
    return rc;
  }

  /// Search the index for the `k` nearest neighbors of `query`. Returns a
  /// list of [VectorResult] (sorted by the C layer).
  List<VectorResult> searchSync(List<double> query, int k) {
    _checkOpen();
    final resultsPtr = calloc<Pointer<vl_result_t>>();
    final nPtr = calloc<Int32>();
    try {
      final rc = WaveDBNative.vectorLayerSearchSync(
          _vl!, query, k, resultsPtr, nPtr);
      if (rc < 0) {
        throw WaveDBException.ioError(
            'vector_layer_search_sync', 'Search failed (rc=$rc)');
      }
      final n = nPtr.value;
      final arr = resultsPtr.value;
      final out = <VectorResult>[];
      if (arr != nullptr && n > 0) {
        for (var i = 0; i < n; i++) {
          final r = (arr + i).ref;
          final idStr = r.id != nullptr
              ? r.id.cast<Utf8>().toDartString()
              : '';
          final metaLen = r.metadataLen;
          Uint8List meta;
          if (r.metadata != nullptr && metaLen > 0) {
            meta = Uint8List.fromList(r.metadata.asTypedList(metaLen));
          } else {
            meta = Uint8List(0);
          }
          out.add(VectorResult(idStr, r.distance, meta));
        }
        WaveDBNative.vectorLayerFreeResults(arr, n);
      }
      return out;
    } finally {
      calloc.free(resultsPtr);
      calloc.free(nPtr);
    }
  }

  /// Delete the vector with id `id`. Returns the C rc (0 on success).
  int deleteSync(String id) {
    _checkOpen();
    final rc = WaveDBNative.vectorLayerDeleteSync(_vl!, id);
    if (rc < 0) {
      throw WaveDBException.ioError(
          'vector_layer_delete_sync', 'Delete failed (rc=$rc)');
    }
    return rc;
  }

  /// train: IVF (re)compute centroids (k-means); SLSH (re)gen projections;
  /// FLAT no-op. Sync-only.
  int train() {
    _checkOpen();
    final rc = WaveDBNative.vectorLayerTrain(_vl!);
    if (rc < 0) {
      throw WaveDBException.ioError(
          'vector_layer_train', 'Train failed (rc=$rc)');
    }
    return rc;
  }

  /// rebuild: rewrite the index's per-vector entries from stored vectors.
  /// Sync-only.
  int rebuild() {
    _checkOpen();
    final rc = WaveDBNative.vectorLayerRebuild(_vl!);
    if (rc < 0) {
      throw WaveDBException.ioError(
          'vector_layer_rebuild', 'Rebuild failed (rc=$rc)');
    }
    return rc;
  }

  /// Number of vectors currently in the index.
  int count() {
    _checkOpen();
    return WaveDBNative.vectorLayerCount(_vl!);
  }

  /// Reconfigure runtime-tunable params. Format tier is immutable after
  /// create — to change index_type/dim/delimiter/ivf_n_clusters/slsh_*,
  /// drop the layer and recreate.
  int reconfigure(VectorRuntime rt) {
    _checkOpen();
    final rtPtr = calloc<vector_layer_runtime_t>();
    try {
      _fillRuntime(rtPtr.ref, rt);
      final rc = WaveDBNative.vectorLayerReconfigure(_vl!, rtPtr);
      if (rc < 0) {
        throw WaveDBException.ioError(
            'vector_layer_reconfigure', 'Reconfigure failed (rc=$rc)');
      }
      return rc;
    } finally {
      calloc.free(rtPtr);
    }
  }

  // ── lifecycle ──

  /// Close the layer and release C resources. Idempotent.
  void dispose() {
    if (_closed || _vl == null) return;
    _closed = true;
    _finalizer.detach(this);
    WaveDBNative.vectorLayerDestroy(_vl!);
    _vl = null;
  }

  /// Alias for [dispose] — mirrors graph_layer.dart's close().
  void close() => dispose();

  void _checkOpen() {
    if (_closed || _vl == null) {
      throw StateError('VectorLayer is closed');
    }
  }
}

// ── C struct fillers ──

Pointer<vector_layer_config_t> _newConfig(VectorFormat fmt, VectorRuntime rt) {
  final cfg = calloc<vector_layer_config_t>();
  _fillFormat(cfg.ref.format, fmt);
  _fillRuntime(cfg.ref.runtime, rt);
  return cfg;
}

void _fillFormat(vector_layer_format_t ref, VectorFormat fmt) {
  // Validate delimiter: must encode to a single byte.
  int delim;
  if (fmt.delimiter.length == 1) {
    final units = fmt.delimiter.codeUnits;
    if (units.length != 1 || units[0] > 255) {
      throw VectorLayerException(
          'delimiter must encode to a single byte, got ${fmt.delimiter}');
    }
    delim = units[0];
  } else if (fmt.delimiter.isEmpty) {
    delim = 0x2f; // '/'
  } else {
    throw VectorLayerException(
        'delimiter must be a single character, got ${fmt.delimiter}');
  }
  ref.indexType = fmt.indexType;
  ref.dim = fmt.dim;
  ref.delimiter = delim;
  ref.distance = fmt.distance;
  ref.ivfNClusters = fmt.ivfNClusters;
  ref.slshLshTables = fmt.slshLshTables;
  ref.slshHashBits = fmt.slshHashBits;
  ref.slshBucketWidth = fmt.slshBucketWidth;
}

void _fillRuntime(vector_layer_runtime_t ref, VectorRuntime rt) {
  ref.topK = rt.topK;
  ref.syncOnly = rt.syncOnly;
  ref.ivfNprobe = rt.ivfNprobe;
  ref.ivfFlatUntil = rt.ivfFlatUntil;
  ref.slshScanRadius = rt.slshScanRadius;
}

File: bindings/nodejs/lib/vector_layer.js
'use strict';

// VectorLayer — Node.js wrapper for the WaveDB vector layer C API.
//
// Mirrors the GraphLayer JS wrapper. Two construction modes:
//   - VectorLayer.openSeparate(dbLocation, indexName, format, runtime)
//       Opens a DEDICATED vector database at dbLocation (no subtree).
//   - VectorLayer.open(indexName, db, format, runtime, subtree?)
//       Opens on an EXISTING WaveDB instance (shared key space). The
//       `db` argument is a WaveDB instance (we extract its underlying
//       database_t* via _getDbPtr). `subtree?` is an optional Subtree
//       instance — when provided, all keys land under that subtree
//       prefix.
//
// Config is split into Format (immutable after create) and Runtime
// (mutable via reconfigure). The C API has no slsh_bidirectional field
// (SLSH always scans bidirectionally) — this binding does NOT expose
// one. Trying to set it in runtimeObj is a no-op (the field is silently
// ignored, matching the C struct).
//
// Sync API is wired directly. Async variants (insert/search/delete)
// return Promises but currently just await the sync variant — the
// promise-based C async API needs a ThreadSafeFunction bridge which is
// a follow-up. JS callers can still benefit from non-blocking behaviour
// by using queueMicrotask / setImmediate wrappers if needed, but the
// underlying call runs synchronously on the main thread.

const native = require('../build/Release/wavedb.node');
const { VectorLayer: VectorLayerNative, VL_INDEX_FLAT, VL_INDEX_IVF, VL_INDEX_SLSH,
        VL_DIST_L2, VL_DIST_COSINE, VL_DIST_DOT } = native;

// Enum constants exposed for ergonomics.
const IndexType = {
  FLAT: VL_INDEX_FLAT,
  IVF:  VL_INDEX_IVF,
  SLSH: VL_INDEX_SLSH,
};
const Distance = {
  L2:     VL_DIST_L2,
  COSINE: VL_DIST_COSINE,
  DOT:    VL_DIST_DOT,
};

class VectorLayer {
  /**
   * Open a DEDICATED vector database at dbLocation.
   *
   * @param {string} dbLocation - Filesystem path for the vector db
   * @param {string} indexName - Index name (namespaces the layer's keys)
   * @param {object} format - Format config (immutable after create)
   * @param {object} runtime - Runtime config (mutable via reconfigure)
   * @returns {VectorLayer}
   */
  static openSeparate(dbLocation, indexName, format, runtime) {
    if (typeof dbLocation !== 'string' || !dbLocation) {
      throw new TypeError('dbLocation string required');
    }
    if (typeof indexName !== 'string' || !indexName) {
      throw new TypeError('indexName string required');
    }
    const fmt = _normalizeFormat(format);
    const rt = _normalizeRuntime(runtime);
    const nativeLayer = new VectorLayerNative(dbLocation, indexName, fmt, rt);
    const vl = new VectorLayer();
    vl._native = nativeLayer;
    vl._closed = false;
    return vl;
  }

  /**
   * Open on an EXISTING WaveDB instance (shared key space).
   *
   * @param {string} indexName - Index name
   * @param {WaveDB} db - A WaveDB instance (we extract its database_t*)
   * @param {object} format - Format config (immutable after create)
   * @param {object} runtime - Runtime config (mutable via reconfigure)
   * @param {Subtree} [subtree] - Optional Subtree (keys land under its prefix)
   * @returns {VectorLayer}
   */
  static open(indexName, db, format, runtime, subtree) {
    if (typeof indexName !== 'string' || !indexName) {
      throw new TypeError('indexName string required');
    }
    if (!db || typeof db._getDbPtr !== 'function') {
      throw new TypeError('db must be a WaveDB instance');
    }
    const dbPtr = db._getDbPtr();
    const fmt = _normalizeFormat(format);
    const rt = _normalizeRuntime(runtime);
    let subtreePtr = null;
    if (subtree && typeof subtree._getPtr === 'function') {
      subtreePtr = subtree._getPtr();
    }
    const args = [indexName, dbPtr, fmt, rt];
    if (subtreePtr !== null) args.push(subtreePtr);
    const nativeLayer = new VectorLayerNative(...args);
    const vl = new VectorLayer();
    vl._native = nativeLayer;
    vl._closed = false;
    return vl;
  }

  // --- Sync operations ---

  /**
   * Insert a vector with an optional metadata buffer.
   * @param {string} id
   * @param {Float32Array} vec
   * @param {Buffer} [metadata]
   */
  insertSync(id, vec, metadata) {
    _checkOpen(this);
    if (!(vec instanceof Float32Array)) {
      throw new TypeError('vec must be a Float32Array');
    }
    if (metadata !== undefined && metadata !== null && !Buffer.isBuffer(metadata)) {
      throw new TypeError('metadata must be a Buffer or null/undefined');
    }
    this._native.insertSync(id, vec, metadata || null);
  }

  /**
   * Insert a batch of vectors atomically.
   * @param {Array<{id: string, vec: Float32Array, metadata?: Buffer}>} items
   */
  insertBatchSync(items) {
    _checkOpen(this);
    if (!Array.isArray(items)) throw new TypeError('items must be an array');
    this._native.insertBatchSync(items);
  }

  /**
   * Search for the k nearest neighbors of `query`.
   * @param {Float32Array} query
   * @param {number} k
   * @returns {Array<{id: string, distance: number, metadata: Buffer|null}>}
   */
  searchSync(query, k) {
    _checkOpen(this);
    if (!(query instanceof Float32Array)) {
      throw new TypeError('query must be a Float32Array');
    }
    return this._native.searchSync(query, k);
  }

  /**
   * Delete a vector by id.
   * @param {string} id
   */
  deleteSync(id) {
    _checkOpen(this);
    this._native.deleteSync(id);
  }

  /**
   * Train the index (IVF: compute centroids; SLSH: generate projections;
   * FLAT: no-op).
   */
  train() {
    _checkOpen(this);
    this._native.train();
  }

  /**
   * Rewrite the index's per-vector entries from the stored vectors.
   */
  rebuild() {
    _checkOpen(this);
    this._native.rebuild();
  }

  /**
   * @returns {number} Number of vectors in the index.
   */
  count() {
    _checkOpen(this);
    return this._native.count();
  }

  /**
   * Reconfigure runtime-tunable params. Format tier is immutable.
   * @param {object} runtime
   */
  reconfigure(runtime) {
    _checkOpen(this);
    this._native.reconfigure(_normalizeRuntime(runtime));
  }

  /**
   * Close the layer and release native resources.
   */
  close() {
    if (this._native && !this._closed) {
      this._closed = true;
      this._native.close();
    }
  }

  get isClosed() {
    return this._closed;
  }

  // --- Async wrappers (deferred: call sync variants, wrap in Promise) ---
  //
  // These return Promises for API parity with the Graph layer, but the
  // underlying call currently runs synchronously on the main thread.
  // A true async bridge (ThreadSafeFunction + C promise_t) is a
  // follow-up — see vector_layer.cc header comment.

  /**
   * Async insert (currently sync-under-the-hood; returns a Promise).
   */
  insert(id, vec, metadata) {
    return Promise.resolve().then(() => {
      this.insertSync(id, vec, metadata);
    });
  }

  /**
   * Async batch insert (currently sync-under-the-hood; returns a Promise).
   */
  insertBatch(items) {
    return Promise.resolve().then(() => {
      this.insertBatchSync(items);
    });
  }

  /**
   * Async search (currently sync-under-the-hood; returns a Promise).
   */
  search(query, k) {
    return Promise.resolve().then(() => this.searchSync(query, k));
  }

  /**
   * Async delete (currently sync-under-the-hood; returns a Promise).
   */
  del(id) {
    return Promise.resolve().then(() => {
      this.deleteSync(id);
    });
  }
}

// --- helpers ---

function _checkOpen(vl) {
  if (vl._closed) throw new Error('VectorLayer is closed');
  if (!vl._native) throw new Error('VectorLayer not initialized');
}

// Normalize a user-supplied format object: ensure index_type / distance
// are integers, drop any unknown fields (e.g. slsh_bidirectional — the
// C API has no such field, so we silently ignore it rather than error,
// matching the C struct's "what you set is what you get" contract).
function _normalizeFormat(format) {
  if (!format || typeof format !== 'object') {
    throw new TypeError('format object required');
  }
  const out = {
    dim: format.dim | 0,
  };
  if (format.index_type !== undefined) {
    out.index_type = format.index_type | 0;
  }
  if (format.delimiter !== undefined) {
    out.delimiter = format.delimiter;
  }
  if (format.distance !== undefined) {
    out.distance = format.distance | 0;
  }
  if (format.ivf_n_clusters !== undefined) {
    out.ivf_n_clusters = format.ivf_n_clusters | 0;
  }
  if (format.slsh_lsh_tables !== undefined) {
    out.slsh_lsh_tables = format.slsh_lsh_tables | 0;
  }
  if (format.slsh_hash_bits !== undefined) {
    out.slsh_hash_bits = format.slsh_hash_bits | 0;
  }
  if (format.slsh_bucket_width !== undefined) {
    out.slsh_bucket_width = +format.slsh_bucket_width;
  }
  return out;
}

// Normalize a user-supplied runtime object. Unknown fields (e.g.
// slsh_bidirectional) are silently dropped.
function _normalizeRuntime(runtime) {
  if (!runtime || typeof runtime !== 'object') {
    throw new TypeError('runtime object required');
  }
  const out = {};
  if (runtime.top_k !== undefined)           out.top_k           = runtime.top_k | 0;
  if (runtime.sync_only !== undefined)       out.sync_only       = runtime.sync_only | 0;
  if (runtime.ivf_nprobe !== undefined)      out.ivf_nprobe      = runtime.ivf_nprobe | 0;
  if (runtime.ivf_flat_until !== undefined)  out.ivf_flat_until  = runtime.ivf_flat_until | 0;
  if (runtime.slsh_scan_radius !== undefined) out.slsh_scan_radius = runtime.slsh_scan_radius | 0;
  return out;
}

module.exports = {
  VectorLayer,
  IndexType,
  Distance,
  // Re-export raw enum values too for callers who want them.
  VL_INDEX_FLAT,
  VL_INDEX_IVF,
  VL_INDEX_SLSH,
  VL_DIST_L2,
  VL_DIST_COSINE,
  VL_DIST_DOT,
};